doc/bmp/bmp.md

```diff
@@ -0,0 +1,373 @@
+ 1) "next_hop"
+ 2) "10.0.0.63"
+ 3) "local_pref"
+ 4) "0"
+ 5) "aggregator"
+ 6) ""
+ 7) "originator_id"
+ 8) ""
+ 9) "origin"
+10) "igp"
+11) "as_path"
+12) " 65100 64600 65534 64627 65505"
+```
+
+
+## 2.5 BMP Agent
+As [2.2 OpenBMP](#22-openbmp) shown, We need to fork and update code in [OpenBMPd](#https://github.com/SNAS/openbmp/tree/master/Server/). OpenBMP supports BMP protocol collecting by openbmpd agent. Thus in this project we will only need openbmpd agent role, and add redis population when monitoring BGP data from BGP container.
+
+Below picture is referenced from [OpenBMPFlow](#https://www.openbmp.org/#openbmp-flow/), refer the part in <span style="color: red;">red</span> circle, which is the daemon we need to update in this porject.
+
+<img src="images/openbmp.png" alt="OPENBMP ARCHITECTURE" width="400">  
+
+
+### Detail workflow
+With source code of openbmpd, we need to update code in message parser and populate redis with specific table required, this should be controlled in runtime by CLI (as more details in below section). Here following the straightforward option, whenever FRR launches TYPE_INIT_MSG to BMP, BMP will just clear redis relevant table to keep the data consistency, since following the TYPE_INIT_MSG, all BGP relevant data will be reSynced from FRR.
+
+<img src="images/bmp_seq.png" alt="BMP brief sequence diagram" width="800">  
+
+### Delay-removal
+There will be data flapping if BGP/FRR connection to BMP is not stable, or even some flapping occurs in BGP side. In order to optimize this scenario handling. Below delay-removal algo could be used internally.
+
+1. When FRR connects to the BMP server with TYPE_INIT_MSG and sends BGP session updates, instead of immediately removing the entire table from Redis, just mark down the timestamp.
+2. Maintain a timer or counter to keep track of the elapsed time since the last update received from FRR.
+3. Set a threshold or predefined delay period for considering a BGP session update as stale or removed. For example, you can define that a BGP session update is considered stale if it has not been updated within a specific time interval (e.g., 3 minutes).
+4. Periodically check the timestamp of the BGP session updates stored in table. If the elapsed time since the last update exceeds the threshold defined in step 3, mark the corresponding entry as removed or delete it from Redis.
+5. Once re-adding the table with the new data sent by FRR arrives, start calculating the dataset and erasing the timer for data refreshed, and let staled data be removed by original loop.
+
+This ensures data consistency and minimizes unnecessary table flapping.
+
+### Config DB schema
+
+Create new config db table as below to contain config list for bmp.
+
+```
+   127.0.0.1:6379[4]> keys FEATURE|bmp*
+   1) "FEATURE|bmp"
+
+```
+
+Create below config items list for enabling and disabling different table.
+
+```
+   127.0.0.1:6379[4]> HGETALL FEATURE|bmp
```

Comment:
Since FRR does support bmp enable/disable as a cmd-line arg (`/usr/lib/frr/bgpd -A 127.0.0.1 -M bmp`), to support per table level enable/disable as shown below, it's better to discuss this with FRR-community and introduce table level enable/disable support via FRR CLI knob, so that zebra/bgpd doesn't have to spend any cycle in sending change notifications for those tables to BMP ; Otherwise zebra/bgpd will always send notifications to BMP for the disabled tables and bmp ignores it later based on this config-db.

Please discuss with FRR community on this, as discussed in the HLD review community meeting (March-12-2024)

Reply:
Sure, currently FRR just post data to BMP following standard BGP protocol, thus neighbor table will be sent in BGP_MSG_OPEN message (#https://datatracker.ietf.org/doc/html/rfc4271#page-13) and all other tables including rib/ls_state... will be sent in BGP_MSG_UPDATE. A better option is to update in FRR side and controlled by the config/cli together, so that no extra data is transmitted between FRR and BMP. Will follow up with FRR community.

Reply:
Thanks. Let's discuss with FRR community and enhance FRR capability to achieve this, as part of bringing in this feature to community-sonic.